kodilla-hibernate/src/test/java/com/kodilla/hibernate/invoice/dao/InvoiceDaoTestSuite.java

```diff
@@ -0,0 +1,124 @@
+package com.kodilla.hibernate.invoice.dao;
+
+import com.kodilla.hibernate.invoice.Invoice;
+import com.kodilla.hibernate.invoice.Item;
+import com.kodilla.hibernate.invoice.Product;
+import org.junit.After;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.test.context.junit4.SpringRunner;
+
+import java.math.BigDecimal;
+import java.util.List;
+
+import static org.junit.Assert.assertEquals;
+
+@RunWith(SpringRunner.class)
+@SpringBootTest
+public class InvoiceDaoTestSuite {
+    @Autowired
+    private InvoiceDao invoiceDao;
+
+    @Autowired
+    private ProductDao productDao;
+
+    @After
+    public void after() {
+        System.out.println("after");
+    }
+
+    @Test
+    public void testSaveNewProducts() {
+        // Given
+        Product product1 = new Product("Bread");
+        Product product2 = new Product("Butter");
+        Product product3 = new Product("Cheese");
+
+        Invoice invoice = new Invoice("F/1/2018");
+        List<Item> items = invoice.getItems();
+        Item item1 = new Item(product1, new BigDecimal("3.20"), 3);
+        items.add(item1);
+        item1.setInvoice(invoice);
+        Item item2 = new Item(product2, new BigDecimal("4.60"), 1);
+        items.add(item2);
+        item2.setInvoice(invoice);
+        Item item3 = new Item(product3, new BigDecimal("5.35"), 2);
+        items.add(item3);
+        item3.setInvoice(invoice);
+
+        // When
+        invoiceDao.save(invoice);
+        int id = invoice.getId();
+        Invoice readInvoice = invoiceDao.findOne(id);
+        List<Item> itemsFromSavedInvoice = invoice.getItems();
+        List<Item> itemsFromReadInvoice = readInvoice.getItems();
+
+        // Then
+        try {
+            assertEquals(invoice, readInvoice);
+            assertEquals(3, items.size());
+            assertEquals(3, itemsFromSavedInvoice.size());
+            assertEquals(3, itemsFromReadInvoice.size());
+            assertEquals(item1, items.get(0));
+            assertEquals(item2, items.get(1));
+            assertEquals(item3, items.get(2));
+            assertEquals(item1, itemsFromSavedInvoice.get(0));
+            assertEquals(item2, itemsFromSavedInvoice.get(1));
+            assertEquals(item3, itemsFromSavedInvoice.get(2));
+            assertEquals(item1, itemsFromReadInvoice.get(0));
+            assertEquals(item2, itemsFromReadInvoice.get(1));
+            assertEquals(item3, itemsFromReadInvoice.get(2));
+        } catch (AssertionError error) {
+            throw error;
+        } finally {
+            invoiceDao.delete(id);
+        }
+    }
+
+    // testSaveReadProducts() przechodzi gdy jest w Item.java line 43: CascadeType.MERGE.
```

Comment:
CascadeType.All aggregates among others CascadeType.PERISIST. In case of this option, when Item is persisted (INSERT INTO ...) , it is seems that it also tries to persist Product - and because object already has id it throws exception. See entity manger methods: https://docs.oracle.com/javaee/7/api/javax/persistence/EntityManager.html
It means that you need to decide if you will pass already stored product or new product which will be persisted when saving Item.